### PR TITLE
notification issue fixed

### PR DESCRIPTION
**Pull request by mebjas**
Signed-off-by: Minhaz A V <minhazav@gmail.com>

Changes
 - `isPermissionDefault()`: `added` a method to check id the current notification preference is `default`.

 - `showToolbar()`: `added` a method that returns bool value corresponding to whether or not we should show the toolbar. It first checks localStorage if the user has clicked on close button of the toolbar, then checks if the status of notification permission isn't default.

- `setEnabled ()`: now uses the `then()` method of `Notification` class to deal with user response to activate notification dialog. Performs necessary actions as per user's response. No changes if the user simply dismisses the dialog.

- `setToolbarHidden ()`: added an extra parameter that defines whether to store the setting persistently or not.

- `setToolbarPersistantHidden ()`: `added` to store the setting to localStorage

### CODE REVIEW

File: src/Notifier.js [hunk 1]
```diff
@@ -123,6 +123,26 @@ var Notifier = {
         return global.Notification.permission == 'granted';
     },
 
+    isPermissionDefault: function() {
+        if (!this.supportsDesktopNotifications()) return false;
+        return global.Notification.permission == 'default';
+    },
+
+    showToolbar: function() {
```
Comment (richvdh) [inline src/Notifier.js]: This doesn't seem to be used anywhere. Is there another PR to come?

`showToolbar` sounds like it will show the toolbar. This should be `shouldShowToolbar` or similar (and it needs a comment saying what it does)

File: src/Notifier.js [hunk 2]
```diff
@@ -133,29 +153,35 @@ var Notifier = {
...
 
         if(enable) {
-            if (!this.havePermission()) {
-                global.Notification.requestPermission(function() {
-                    if (callback) {
-                        callback();
+            // Case when we do not have the permission as 'granted'
+            if (this.isPermissionDefault()) {
+                // Attempt to get permission from user
+                var _this = this;
```
Comment (richvdh) [inline src/Notifier.js]: we normally use `self` for maintaining a reference to `this.

File: src/Notifier.js [hunk 3]
```diff
@@ -133,29 +153,35 @@ var Notifier = {
...
+                global.Notification.requestPermission().then(function(result) {
+                    if (result === 'denied') {
                         dis.dispatch({
```
Comment (richvdh) [inline src/Notifier.js]: The `denied` case doesn't change anything, so I don't think we need to dispatch the `notifier_enabled` case here

File: src/Notifier.js [hunk 4]
```diff
@@ -133,29 +153,35 @@ var Notifier = {
...
                             action: "notifier_enabled",
-                            value: true
+                            value: false
                         });
+                        _this.setToolbarHidden(true);
```
Comment (richvdh) [inline src/Notifier.js]: I also don't think it's correct to hide the toolbar when the user clicks 'block' on the dialog.

File: src/Notifier.js [hunk 5]
```diff
@@ -133,29 +153,35 @@ var Notifier = {
...
         if(enable) {
-            if (!this.havePermission()) {
-                global.Notification.requestPermission(function() {
-                    if (callback) {
-                        callback();
+            // Case when we do not have the permission as 'granted'
+            if (this.isPermissionDefault()) {
+                // Attempt to get permission from user
+                var _this = this;
+                global.Notification.requestPermission().then(function(result) {
```
Comment (richvdh) [inline src/Notifier.js]: I don't think we can rely on our users having browsers which support the promise-returning version of `requestPermission`. https://developer.mozilla.org/en-US/docs/Web/API/Notification/requestPermission suggests that Chrome doesn't support it at all. You'll need to stick with the callback version.

File: src/Notifier.js [hunk 6]
```diff
@@ -133,29 +153,35 @@ var Notifier = {
         }
 
         if(enable) {
-            if (!this.havePermission()) {
-                global.Notification.requestPermission(function() {
-                    if (callback) {
-                        callback();
+            // Case when we do not have the permission as 'granted'
+            if (this.isPermissionDefault()) {
```
Comment (richvdh) [inline src/Notifier.js]: I don't think there's any need for this test. `global.Notification.requestPermission()` will do nothing (except give result==granted) if we already have permission.

File: src/Notifier.js [hunk 7]
```diff
@@ -192,12 +216,22 @@ var Notifier = {
         return enabled === 'true';
     },
 
-    setToolbarHidden: function(hidden) {
+     setToolbarHidden: function(hidden, persistent) {
```
Comment (richvdh) [inline src/Notifier.js]: indentation has gone wonky here

Comment (richvdh) [inline src/Notifier.js]: Also, when do you expect this to be called with `persistent` == `true`?

File: src/Notifier.js [hunk 8]
```diff
@@ -192,12 +216,22 @@ var Notifier = {
...
         dis.dispatch({
             action: "notifier_enabled",
             value: this.isEnabled()
         });
+
+        if (persistent === true)
+            this.setToolbarPersistantHidden();
+    },
+
+    setToolbarPersistantHidden: function() {
```
Comment (richvdh) [inline src/Notifier.js]: s/Persistant/Persistent/, but to be honest you might as well inline this.

File: src/Notifier.js [hunk 9]
```diff
@@ -192,12 +216,22 @@ var Notifier = {
...
 
-    setToolbarHidden: function(hidden) {
+     setToolbarHidden: function(hidden, persistent) {
         this.toolbarHidden = hidden;
         dis.dispatch({
             action: "notifier_enabled",
             value: this.isEnabled()
         });
+
+        if (persistent === true)
```
Comment (richvdh) [inline src/Notifier.js]: I'd just make this `if (persistent)`. Also please use `{}` even for one-line if statements.

**Comment by richvdh** [comment]
For the record: this relates to https://github.com/vector-im/vector-web/issues/1095 and https://github.com/vector-im/vector-web/issues/951

**Comment by richvdh** [comment]
@mebjas: Thanks very much for the PR. Generally it looks good, though maybe a bit more complicated than I expected. I've made a few comments which might also help simplify things a bit.

File: src/Notifier.js (hunk 1 above)
Comment (mebjas) [inline src/Notifier.js]: It's used by client to know weather or not to show the toolbar, I'll change it to `shouldShowToolbar()` though
https://github.com/vector-im/vector-web/pull/1237

File: src/Notifier.js (hunk 6 above)
Comment (mebjas) [inline src/Notifier.js]: Well yeah normally it would never lead to this, as if `global.Notification.permission` != `default` the toolbar wouldn't even be shown.
But still in SDK shouldn't we have the case where we hide the toolbar in case the  `global.Notification.permission` is already `granted` or `denied`

File: src/Notifier.js (hunk 2 above)
Comment (mebjas) [inline src/Notifier.js]: cool will do

File: src/Notifier.js (hunk 5 above)
Comment (mebjas) [inline src/Notifier.js]: :O it supported on chrome, but yeah we need to have backward compatibility.

File: src/Notifier.js (hunk 4 above)
Comment (mebjas) [inline src/Notifier.js]: Once the user has blocked, only way is to manually change the settings from url bar. We cannot raise the dialog anyways after that

File: src/Notifier.js (hunk 7 above)
Comment (mebjas) [inline src/Notifier.js]: It's called when user attempts to close the toolbar. This setting should be persistent. Currently the toolbar keep showing up every reload.
`setToolbarPersistantHidden()` set the information in localStorage

File: src/Notifier.js (hunk 8 above)
Comment (mebjas) [inline src/Notifier.js]: w/o a function?

Comment (richvdh) [inline src/Notifier.js]: yes. it's only two lines, and only used in one place. Having a separate function doesn't do much for code clarity.

File: src/Notifier.js (hunk 4 above)
Comment (richvdh) [inline src/Notifier.js]: If the "You are not receiving desktop notifications" bar disappears, it gives the impression that now you will be receiving desktop notifications.

You're right that clicking on the "Enable them now" link won't work, but maybe we should deal with that a different way. (We could hide 'Enable them now', or show a modal dialog giving the user tips on how to enable notifications). Why don't you leave the toolbar shown as it is for now, and we can consider whether to handle this better in a separate PR.

File: src/Notifier.js (hunk 1 above)
Comment (richvdh) [inline src/Notifier.js]: Thanks. Sorry I missed the other PR. It's useful to say something about the other PR in the PR comments!

File: src/Notifier.js (hunk 7 above)
Comment (richvdh) [inline src/Notifier.js]: ok. I suggest you make `true` the default for persistent (and set it to `false` where you call it from `setEnabled`), so that you don't need to change `MatrixToolbar`.

Comment (mebjas) [inline src/Notifier.js]: Cool would make the changes

File: src/Notifier.js [hunk 10]
```diff
@@ -133,29 +153,35 @@ var Notifier = {
...
         if(enable) {
-            if (!this.havePermission()) {
-                global.Notification.requestPermission(function() {
-                    if (callback) {
-                        callback();
+            // Case when we do not have the permission as 'granted'
+            if (this.isPermissionDefault()) {
+                // Attempt to get permission from user
+                var self = this;
+                global.Notification.requestPermission().then(function(result) {
```
Comment (mebjas) [inline src/Notifier.js]: So I shall shift this to callback one? I think that MDN doc is little updated, as this is supported in chrome now.

File: src/Notifier.js [hunk 11]
```diff
@@ -123,6 +123,26 @@ var Notifier = {
...
+            if (global.localStorage.getItem('notifications_hidden') === 'true')
+                return false;
+        }
+
+        // Check if permission is granted by any chance.
+        if (this.havePermission()) return false;
+
+        // means the permission is blocked
+        if (!this.isPermissionDefault()) return false;
+        return true;
```
Comment (mebjas) [inline src/Notifier.js]: @richvdh Tell me what do you think about showing a message if the user has blocked notif request **& not yet attempted to close the toolbar** - if yes, I'll change this method to return enums - 'show', 'hide', 'blocked' and the vector will take decision accordingly.

File: src/Notifier.js (hunk 10 above)
Comment (richvdh) [inline src/Notifier.js]: Yes, I think you might as well use the callback flow.

Obviously the MDN doc is outdated, but we don't know which version of Chrome introduced the promise; we don't know if it works in Safari; and using the promise doesn't buy us anything.

File: src/Notifier.js [hunk 12]
```diff
@@ -192,12 +218,23 @@ var Notifier = {
...
             action: "notifier_enabled",
             value: this.isEnabled()
         });
+
+        if (persistent) {
+            this.setToolbarPersistantHidden();
+        }
+    },
+
+    setToolbarPersistantHidden: function() {
```
Comment (richvdh) [inline src/Notifier.js]: still needs inlining

File: src/components/structures/MatrixChat.js [hunk 13]
```diff
@@ -1046,7 +1046,7 @@ module.exports = React.createClass({
             if (MatrixClientPeg.get().isGuest()) {
                 topBar = <GuestWarningBar />;
             }
-            else if (Notifier.supportsDesktopNotifications() && !Notifier.isEnabled() && !Notifier.isToolbarHidden()) {
+            else if (Notifier.supportsDesktopNotifications() && Notifier.shouldShowToolbar()) {
```
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: I don't think this change is necessary. It's clearer to keep `isEnabled` and `isToolbarHidden` separate imho

File: src/Notifier.js [hunk 14]
```diff
@@ -192,12 +218,23 @@ var Notifier = {
...
+        // update the info to localStorage
+        if (global.localStorage) {
+            global.localStorage.setItem('notifications_hidden', 'true');
+        }
     },
 
     isToolbarHidden: function() {
```
Comment (richvdh) [inline src/Notifier.js]: `isToolbarHidden` needs to check the localStorage setting

File: src/Notifier.js [hunk 15]
```diff
@@ -192,12 +218,23 @@ var Notifier = {
         return enabled === 'true';
     },
 
-    setToolbarHidden: function(hidden) {
+    setToolbarHidden: function(hidden, persistent = true) {
```
Comment (richvdh) [inline src/Notifier.js]: just fyi: you may know this already but default parameters are a feature of ES6, which means that we can't use them in some browsers. But it's ok to use in react-sdk because babel translates it for us.

File: src/Notifier.js (hunk 6 above)
Comment (richvdh) [inline src/Notifier.js]: I don't understand your answer.

`setEnabled` is not just used from the toolbar; it is also used from the settings dialog.

But the point is, you do not need to test for `isPermissionDefault` here - just go straight into requestPermission, which will give the same result.

File: src/Notifier.js [hunk 16]
```diff
@@ -133,29 +155,35 @@ var Notifier = {
...
-                global.Notification.requestPermission(function() {
-                    if (callback) {
-                        callback();
+            // Case when we do not have the permission as 'granted'
+            if (this.isPermissionDefault()) {
+                // Attempt to get permission from user
+                var self = this;
+                global.Notification.requestPermission(function(result) {
+                    if (result === 'denied') {
                         dis.dispatch({
```
Comment (richvdh) [inline src/Notifier.js]: still don't think it is necessary to raise an event here

File: src/Notifier.js [hunk 17]
```diff
@@ -133,29 +155,35 @@ var Notifier = {
...
                             action: "notifier_enabled",
-                            value: true
+                            value: false
                         });
+                        self.setToolbarHidden(true, false);
```
Comment (richvdh) [inline src/Notifier.js]: still don't think it is correct to hide the toolbar here

**Comment by richvdh** [comment]
@mebjas: better, but still quite a lot of the issues we discussed remain!

File: src/Notifier.js (hunk 6 above)
Comment (mebjas) [inline src/Notifier.js]: yeah that makes more sense :)

File: src/Notifier.js (hunk 17 above)
Comment (mebjas) [inline src/Notifier.js]: Should atleast change the message or remove the link?

Comment (richvdh) [inline src/Notifier.js]: I'm running out of ways to say this:  Leave it alone for now, and make it a separate PR.

File: src/Notifier.js (hunk 14 above)
Comment (mebjas) [inline src/Notifier.js]: Or change it in `start()` method, where it is defined to be `false`

**Comment by mebjas** [comment]
@richvdh why are we keeping `notifications_enabled` in `localStorage` at all?

File: src/Notifier.js (hunk 17 above)
Comment (mebjas) [inline src/Notifier.js]: Ok, no actions if denied!

**Comment by richvdh** [comment]
We might have permission from the browser to show notifications, but the user has disabled the setting in Vector. That is what `notifications_enabled` is about.

File: src/Notifier.js [hunk 18]
```diff
@@ -133,29 +133,24 @@ var Notifier = {
...
-                        });
-                    }
-                });
-            }
-
-            if (!global.localStorage) return;
-            global.localStorage.setItem('notifications_enabled', 'true');
-
-            if (this.havePermission) {
+            // Case when we do not have the permission as 'granted'
```
Comment (richvdh) [inline src/Notifier.js]: This comment is no longer helpful

File: src/Notifier.js [hunk 19]
```diff
@@ -133,29 +133,24 @@ var Notifier = {
...
+            // Attempt to get permission from user
+            global.Notification.requestPermission(function(result) {
+                if (result !== 'granted') {
+                    // The permission request was dismissed or denied
+                    return;
+                }
+
+                if (callback) callback();
```
Comment (richvdh) [inline src/Notifier.js]: I think you need to update the localStorage setting before you run the callback and dispatch the `notifier_enabled` setting, otherwise things will get confused.

**Comment by richvdh** [comment]
Much better :). Nearly there :)

File: src/Notifier.js (hunk 19 above)
Comment (mebjas) [inline src/Notifier.js]: localStorage setting as in for `notifications_enabled` right?

**Comment by richvdh** [comment]
This is looking really good now. I think there is one remaining problem though: Once you dismiss the banner, it will never, ever come back again. Suppose:
* I don't want to enable notifications so I dismiss the banner
* Later I decide I do want notifications after all so I enable them via the settings dialog
* Later still I turn them off again.
At this point I think it makes sense if we show the banner again.

My suggestion would be that you call `setToolbarHidden(false)` from `setEnabled` if `enabled` is true.

**Comment by mebjas** [comment]
@richvdh ?

**Comment by richvdh** [comment]
looks good, thank you!
